pcsx2/DEV9/pcap_io.cpp

```diff
@@ -35,65 +35,95 @@
 #define PCAP_NETMASK_UNKNOWN 0xffffffff
 #endif
 
-#ifdef _WIN32
-#define mac_address char*
-#else
 pcap_t* adhandle;
-pcap_dumper_t* dump_pcap;
+pcap_dumper_t* dump_pcap = nullptr;
 char errbuf[PCAP_ERRBUF_SIZE];
-mac_address broadcast_mac = {0xff, 0xff, 0xff, 0xff, 0xff, 0xff};
-#endif
+
 int pcap_io_running = 0;
+bool pcap_io_switched;
+
 extern u8 eeprom[];
 
 char namebuff[256];
 
+ip_address ps2_ip;
+mac_address ps2_mac;
+mac_address host_mac;
 
-
-// Fetches the MAC address and prints it
-int GetMACAddress(char* adapter, mac_address* addr)
-{
-	int retval = 0;
 #ifdef _WIN32
-	static IP_ADAPTER_ADDRESSES AdapterInfo[128];
+//IP_ADAPTER_ADDRESSES is a structure that contains ptrs to data in other regions
+//of the buffer, se we need to return both so the caller can free the buffer
+//after it's finished reading the needed data from IP_ADAPTER_ADDRESSES
+bool GetWin32Adapter(const char* name, PIP_ADAPTER_ADDRESSES adapter, PIP_ADAPTER_ADDRESSES* buffer)
+{
+	const int guidindex = strlen("\\Device\\NPF_");
+
+	int neededSize = 128;
+	PIP_ADAPTER_ADDRESSES AdapterInfo = new IP_ADAPTER_ADDRESSES[neededSize];
+	ULONG dwBufLen = sizeof(IP_ADAPTER_ADDRESSES) * neededSize;
 
-	static PIP_ADAPTER_ADDRESSES pAdapterInfo;
-	ULONG dwBufLen = sizeof(AdapterInfo);
+	PIP_ADAPTER_ADDRESSES pAdapterInfo;
 
 	DWORD dwStatus = GetAdaptersAddresses(
 		AF_UNSPEC,
 		GAA_FLAG_INCLUDE_PREFIX,
 		NULL,
 		AdapterInfo,
 		&dwBufLen);
+
+	if (dwStatus == ERROR_BUFFER_OVERFLOW)
+	{
+		DevCon.WriteLn("GetWin32Adapter() buffer too small, resizing");
+		delete[] AdapterInfo;
+		neededSize = dwBufLen / sizeof(IP_ADAPTER_ADDRESSES) + 1;
+		AdapterInfo = new IP_ADAPTER_ADDRESSES[neededSize];
+		dwBufLen = sizeof(IP_ADAPTER_ADDRESSES) * neededSize;
+		DevCon.WriteLn("New size %i", neededSize);
+
+		DWORD dwStatus = GetAdaptersAddresses(
+			AF_UNSPEC,
+			GAA_FLAG_INCLUDE_PREFIX,
+			NULL,
+			AdapterInfo,
+			&dwBufLen);
+	}
 	if (dwStatus != ERROR_SUCCESS)
-		return 0;
+		return false;
 
 	pAdapterInfo = AdapterInfo;
 
-	char adapter_desc[128] = "";
-
-	// Must get friendly description from the cryptic adapter name
-	for (int ii = 0; ii < pcap_io_get_dev_num(); ii++)
-		if (0 == strcmp(pcap_io_get_dev_name(ii), adapter))
-		{
-			strcpy(adapter_desc, pcap_io_get_dev_desc(ii));
-			break;
-		}
-
-	wchar_t wadapter[128];
-	std::mbstowcs(wadapter, adapter_desc, 128);
-
 	do
 	{
-		if (0 == wcscmp(pAdapterInfo->Description, wadapter))
+		if (0 == strcmp(pAdapterInfo->AdapterName, &name[guidindex]))
 		{
-			memcpy(addr, pAdapterInfo->PhysicalAddress, 6);
-			return 1;
+			*adapter = *pAdapterInfo;
+			*buffer = AdapterInfo;
+			return true;
 		}
 
 		pAdapterInfo = pAdapterInfo->Next;
 	} while (pAdapterInfo);
+
+	delete[] AdapterInfo;
+	return false;
+}
+#endif
+
+// Fetches the MAC address and prints it
+int GetMACAddress(char* adapter, mac_address* addr)
+{
+	int retval = 0;
+#ifdef _WIN32
+	IP_ADAPTER_ADDRESSES adapterInfo;
+	PIP_ADAPTER_ADDRESSES buffer;
```

Comment:
can't you use an object that has an automatic lifetime instead of having to call delete[] each time?

Reply:
Like std::vector? can do

Reply:
Ended up using a std::unique_ptr instead